content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+- `Kind: Redis` → `chart: redis`
+- `Kind: VirtualMachine` → `chart: virtual-machine`
+
+Moreover, we don't want to have to add a new type to codegen and recompile our extension api-server
+every time we add a new chart for it to start being served.
+The update should be done dynamicly or provided via a `ConfigMap` by the administrator.
+
+### Two-Way conversion
+
+Currently, we already have integrations and a dashboard that continue to use `HelmRelease` resources.
+At this stage, we didn't want to lose the ability to support this API. Considering that we're simply
+translating one resource into another, support is maintained and it works both ways.
+If you create a `HelmRelease`, you'll get a custom resource in Kubernetes, and if you create a
+custom resource in Kubernetes, it will also be available as a `HelmRelease`.
+
+We don't have any additional controllers that synchronize state between these resources.
+All requests to resources in our extension api-server are transparently proxied to `HelmRelease` and vice versa.
+This eliminates intermediate states and the need to write controllers and synchronization logic.
+
+## Implementation
+
+To implement the Aggregation API, you may consider two projects:
+
+- [apiserver-builder](https://github.com/kubernetes-sigs/apiserver-builder-alpha):
+  Currently in alpha and hasn't been updated for two years. It works like kubebuilder,
+  providing a framework for creating an extension api-server, allowing you to sequentially create
+  a project structure and generate code for your resources.
+- [sample-apiserver](https://github.com/kubernetes/sample-apiserver):
+  A ready-made example of an implemented api-server based on official Kubernetes libraries,
+  which you can use as a foundation for your project.
+
+For practical reasons, we chose the second project. Here's what we needed to do:
+
+### Disable etcd support
+
+In our case, we don't need it since all resources are stored directly in the Kubernetes API.
+
+You can disable etcd options by passing nil to RecommendedOptions.Etcd:
+
+- [Disabling etcd options](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/cmd/server/start.go#L70)
+
+### Generate a common resource type
+
+We called it `Application`, and it looks like this:
+
+- [Application type definition](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/apis/apps/v1alpha1/types.go)
+
+This is a generic type used for any application type, and its handling logic is the same for all charts.
+
+### Configure configuration loading
+
+Since we want to configure our extension api-server via a config file, we formed the config structure in Go:
+
+- [Config type definition](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/config/config.go)
+
+We also modified the resource registration logic so that the resources we create are registered in scheme with different Kind values:
```

Comment:
```suggestion
We also modified the resource registration logic so that the resources we create are registered in scheme with different `Kind` values:
```
?